Lib/test/test_asyncio/test_timeouts.py

```diff
@@ -0,0 +1,242 @@
+                    cancel = True
+                    raise
+                except TimeoutError:
+                    self.fail(
+                        "The only topmost timed out context manager "
+                        "raises TimeoutError"
+                    )
+        self.assertTrue(cancel)
+        self.assertTrue(cm1.expired())
+        self.assertTrue(cm2.expired())
+
+    async def test_waiter_cancelled(self):
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0.01):
+                with self.assertRaises(asyncio.CancelledError):
+                    await asyncio.sleep(10)
+
+    async def test_timeout_not_called(self):
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        async with asyncio.timeout(10) as cm:
+            await asyncio.sleep(0.01)
+        t1 = loop.time()
+
+        self.assertFalse(cm.expired())
+        # finised fast. Very busy CI box requires high enough limit,
+        # that's why 0.01 cannot be used
+        self.assertLess(t1-t0, 2)
+        self.assertGreater(cm.when(), t1)
+
+    async def test_timeout_disabled(self):
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        async with asyncio.timeout(None) as cm:
+            await asyncio.sleep(0.01)
+        t1 = loop.time()
+
+        self.assertFalse(cm.expired())
+        self.assertIsNone(cm.when())
+        # finised fast. Very busy CI box requires high enough limit,
+        # that's why 0.01 cannot be used
+        self.assertLess(t1-t0, 2)
+
+    async def test_timeout_at_disabled(self):
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        async with asyncio.timeout(None) as cm:
+            await asyncio.sleep(0.01)
+        t1 = loop.time()
+
+        self.assertFalse(cm.expired())
+        self.assertIsNone(cm.when())
+        # finised fast. Very busy CI box requires high enough limit,
+        # that's why 0.01 cannot be used
+        self.assertLess(t1-t0, 2)
+
+    async def test_timeout_zero(self):
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0) as cm:
+                await asyncio.sleep(10)
+        t1 = loop.time()
+        self.assertTrue(cm.expired())
+        # finised fast. Very busy CI box requires high enough limit,
+        # that's why 0.01 cannot be used
+        self.assertLess(t1-t0, 2)
+        self.assertTrue(t0 <= cm.when() <= t1)
+
+    async def test_foreign_exception_passed(self):
+        with self.assertRaises(KeyError):
+            async with asyncio.timeout(0.01) as cm:
+                raise KeyError
+        self.assertFalse(cm.expired())
+
+    async def test_foreign_cancel_doesnt_timeout_if_not_expired(self):
+        with self.assertRaises(asyncio.CancelledError):
+            async with asyncio.timeout(10) as cm:
+                raise asyncio.CancelledError
```

Comment:
This test would be more convincing if it used an actual cancel() call.
```suggestion
                asyncio.current_task().cancel()
                await asyncio.sleep(0.01)
```

Reply:
Done